Feature/fat32 open

DrDeano opened the pull request:
Needed to return a pointer to the file system as `@fieldParentPtr(Fat32Self, "instance", fs.instance)` doesn't work as the the pointer to `instance` is on the stack, then goes out of scope when the init is returned. So will be using free stack space.
Part of #216

SamTebbs33 commented on mkfat32.zig:
```diff
@@ -557,13 +557,16 @@ pub const Fat32 = struct {
+    ///                             parameters of a FAT32 image.
+    ///     IN stream: anytype    - The stream to create a new FAT32 image. This stream must
+    ///                             support reader(), writer() and seekableStream() interfaces.
+    ///     IN quick_format: bool - Weather to completely zero the stream initially or zero just
```
s/weather/whether/

SamTebbs33 commented on src/kernel/filesystem/fat32.zig:
```diff
@@ -1195,16 +1216,56 @@ pub fn Fat32FS(comptime StreamType: type) type {
-        fn openImpl(fs: *const vfs.FileSystem, dir: *const vfs.DirNode, name: []const u8) (Allocator.Error || ReadError || SeekError || vfs.Error)!*vfs.Node {
+        fn openImpl(fs: *const vfs.FileSystem, dir: *const vfs.DirNode, name: []const u8) (Allocator.Error || vfs.Error)!*vfs.Node {
             const self = @fieldParentPtr(Fat32Self, "instance", fs.instance);
+            // TODO: Cache the files in this dir, so when opening, don't have to iterator the directory every time
```
:+1: We should definitely add some fs caching at some point as operations will be slow otherwise

SamTebbs33 commented on src/kernel/filesystem/fat32.zig:
```diff
@@ -1195,16 +1216,56 @@ pub fn Fat32FS(comptime StreamType: type) type {
+                },
+            }) |entry| {
+                defer entry.deinit();
+                // File name compare is case insensitive
```
I think file names should be case sensitive, just like on Linux.

DrDeano replied:
FAT32 specifically says that they are case insensitive. but Ext2/3/4 is case sensitive so depends on the filesystem

SamTebbs33 replied:
Oh I see, I didn't realise this sort of thing was part of the spec.

SamTebbs33 commented on src/kernel/filesystem/fat32.zig:
```diff
@@ -1247,6 +1308,7 @@ pub fn Fat32FS(comptime StreamType: type) type {
             self.opened_files.deinit();
             self.allocator.destroy(self.root_node.node);
             self.allocator.destroy(self.fs);
+            self.allocator.destroy(self);
```
It would be good to specify that this function destroys the pointer as well. Alternatively you could rename `init` to `create` and `deinit` to `destroy` to follow the convention from other structs.

I think option two is best.

DrDeano replied:
I think it should still be `init` but have added to the doc comment.

SamTebbs33 replied:
How come you kept it as `init`?

DrDeano replied:
We haven't really been using `create` for initialising structs (apart from task). So would have to change all the inits to creates. Also, I don't really know what the semantic difference is, aren't they basically the same meaning anyway?

SamTebbs33 replied:
I thought the difference between `create` and `init` was that the former allocates on the heap and the latter allocates on the stack. We wouldn't have to change all `init`s to `create`s because most allocate on the stack rather than heap (which is the preferred way because it's more flexible). I personally think that we should keep this convention here to make it very clear if the result is stack allocated or heap allocated, as otherwise we could run into seg-faults and double-frees because we forget if a `deinit` function frees the pointer or not.

DrDeano replied:
That makes sense :)

SamTebbs33 commented on src/kernel/filesystem/fat32.zig:
```diff
@@ -1927,18 +1988,52 @@ test "Fat32FS.write" {
+
+    vfs.setRoot(test_fs.root_node.node);
+
+    const file = try vfs.openFile("/ramdisk_test1.txt", .NO_CREATION);
```
Does `file` need to be checked at all? Such as it pointing to the correct file entry (like how the test fs checks that it is the same as what is in the file tree)?

DrDeano replied:
can add extra tests

SamTebbs33 reviewed (approved):
Boom boom

DrDeano committed:
Add a quick format option for mkfat32


Added open implementation

Needed to return a pointer to the file system as `@fieldParentPtr(Fat32Self, "instance", fs.instance)` doesn't work as the the pointer to `instance` is on the stack, then goes out of scope when the init is returned. So will be using free stack space.
Part of #216

Spelling, doc comments and extra test for open file


init -> create, deinit -> destroy
